### PR TITLE
[E2E refactor - 2] change e2e to create and run on a kind cluster per Test* function

### PR DESCRIPTION
**Pull request by pintohutch**
I decided to break out the kind E2E test refactor PR https://github.com/GoogleCloudPlatform/prometheus-engine/pull/738 into smaller, digestible PRs for reviewing.

Note: because of the nature of the change, presubmits (i.e. Github Actions) may fail until the final PR is merged.

This is the second one, where we change the `e2e` make target to provision and run e2e tests against a kind cluster per `Test*` function.

We also build the go binaries and deploy them to a local image registry, which the kind clusters use to run the images from. This way, when we change code, that change is reflected in the components running in the cluster, without relying on published gke images.

Finally, we introduce using `GOOGLE_APPLICATION_CREDENTIALS` as a path to a service account file. If specified, the e2e tests will be run against the Cloud Monitoring backend to assert metric data is properly collected.

### CODE REVIEW

**Review by TheSpiritXIII — CHANGES_REQUESTED** [review]
Thanks! Leaving some questions to start and help move this along. :)

**Review by pintohutch — COMMENTED** [review]
Thanks for the feedback! Made a few changes and responded in-line to your comments.

**Comment by pintohutch** [comment]
Friendly ping here @TheSpiritXIII

File: Makefile [hunk 1]
```diff
@@ -148,31 +176,43 @@ else
 	@echo "Secret not provided, skipping!"
 endif
 
+# TODO(pintohutch): re-enable e2e testing against an existing K8s cluster
+# (e.g. GKE cluster) without relying on a fresh kind cluster.
+# This was the previous behavior of running `NO_DOCKER=1 make e2e`. But maybe
```
Comment (bwplotka) [inline Makefile:181]: 👍🏽

File: Makefile [hunk 2]
```diff
@@ -149,30 +176,38 @@ else
...
+	@echo ">> running kind tests"
+# TODO(pintohutch): handle SIGINTs gracefully. For now a:
+# docker stop $(docker ps -a -q) && docker container prune -f
+# will do the trick.
+	echo $(TEST_RUN) | tr ' ' '\n' | xargs -I {} -P$(KIND_PARALLEL) \
+		docker run \
+		--env GOOGLE_APPLICATION_CREDENTIALS="$(LOCAL_CREDENTIALS)" \
+		--env PROJECT_ID="$(PROJECT_ID)" \
+		--env GMP_LOCATION="$(GMP_LOCATION)" \
+		--env BINARIES="$^" \
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply Makefile:208]: I need to agree with some points we discussed with @TheSpiritXIII a week ago. 

Generally the main points are:

A) Is moving out of Go test framework and replace with more bash a good direction?
B) Do we want to double down on running all scripts in docker containers?

Both points have pros & cons, and there might be some preferences, but sounds like those PRs goes into more bash and more scripts run in docker.

Not sure how we want to get into some decision on those points, we might want to focused discussion and decision. 

I am happy to provide my opinion, but not sure if helpful at this comment. TL;DR  I am generally in favor of relying more on Go, less on Bash and using containers only when it's necessary.

For example:

A) It feels the same what's here in this section can be achieved with t.Parallel with go test. You can use go test flags to adjust how many CPUs, how to shard it further with GitHub actions and some bash ([1](https://github.com/thanos-io/thanos/blob/main/.github/workflows/go.yaml#L97), [2](https://github.com/thanos-io/thanos/blob/main/Makefile#L336)).

B) I understand the perfection of running everything in docker for ultimate reliability. But are we building those dev tools for 10k engineers on various devices from RaspberryPi, through Windows servers, to Mainframe or for 5 engineers with only Mac and one Linux OS? (:  My preference is to run containers for kind. Then it's useful to build docker containers in another container so we can assure multi target builds (although could be avoided perhaps). But then everything else could be in Makefile + bingo (maybe I should just do it, so the check if visible, we discussed we wanted this). The environment reliability will be risker yes, occasionally we will see something working locally but not on CI, but isn't this a bit similar to fighting with script working locally vs our dockerized env locally?

Generally using docker for scripts it's not only slower (maybe it does not matter for CI), but more importantly it's massive pain to change it (you have to pass extra args, rebuild environment image (how, where it is), pass volumes) and to debug it. 

But yea, it's just preference, both ways would work, but we won't be productive if we all don't accept in some form where we want be (:

[the remark this answers is not in the document]
Comment (bwplotka) [reply Makefile:208]: So we have two choices. 

A) Try to address those changes within this stream of PRs
B) Merge all and address later. We already run things in docker, so we might as well be consistent for now and change ASAP later. The parallelization changes are more tricky to revert if needed.

I think I would be fine with B to respect @pintohutch time, maybe modulo parallelization changes 🤔 (use Go test for that for now), we could consider that decision now.

**Review by bwplotka — COMMENTED** [review]
Epic, some comments to move discussion forward. It would be epic to decide one way or another before this will get stale 🤗 

Thanks for epic work @pintohutch and solid review @TheSpiritXIII  💪🏽

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (pintohutch) [reply Makefile:208]: Thanks for the discussion @bwplotka. @TheSpiritXIII and I had a conversation about this yesterday actually.

tl;dr - IIUC the biggest thing we lose with this chain of PRs is lack of interactive debugging (i.e. in IDE) without go test and operator running locally. However, we should be able to add that as a separate "mode" on top of these changes without breaking anything.

There's no more relying on bash vs go here. It's the exact same pattern as before (no-op) with a few additions and changes in philosophy:
1. Addition: operator runs in-cluster.
1. Addition: we `docker build` images from HEAD to be used in the cluster as part of the e2e test (calling `docker` in `go test` feels weird).
1. Change (biggest): we run each `TestXXX` on a fresh `kind` cluster instead of re-using an existing one. The reasons there are many and happy to discuss synchronously. However, I did leave a TODO to reuse an existing cluster if that's desired. Should be a pretty easy improvement.
1. Change: Parallelization using bash over go. This is a little hacky, I admit. I think most of the time developers won't use it locally. I'm happy to rip it out entirely (in favor of sequential tests), as these tests will run in parallel (and much faster) on Github as part of our CI. But I think it's fairly benign.
3. No-change: running scripts in docker. We've always done that. I actually tried out a [mode](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/d776226ceddbce346d2a44334955c4b9724baeaa/Makefile#L192) where we didn't do that in the other branch. I am happy to reintroduce that, but you have to be much more careful with cleanup, bookkeeping, etc. I found it easier to eat the extra 5-10% runtime and have less code (when runtime delta is maybe 1-2mins). locally. When I tried on Github Actions, it maybe made things 30s faster. Happy to improve here.

[the remark this answers is not in the document]
Comment (bwplotka) [reply Makefile:208]: Nice, thanks for listing out.

I think some of those things are an improvement. For instance I personally love 1 and 2. It is super important and clearly we missed some cases because of that.

However I think 3, 4, 5 are tricky. They might be improvement on some level, but they add more problems on others. I think I am not fan of those. What is generally motivation for those? It does not help with reliability of our test (based on data -- we never had a case of false positive (e.g. "successful test of buggy because of cluster state was reused") or false negative). To me 3 and 4 aims to reduce code complexity then? For the framework (e.g. that bookkeeping)? What if the team prefers more complex framework, but more flexibility on parallelism? 5 is a problem, as it's just harder to iterate on test framework (e.g. add another flag etc).

To me the priority on e2e tests generally:

A. Reliability e.g. including testing RBAC + webhooks
B. Feedback time -- so test time (imagine we will have more components e.g. Otel, more test cases etc), it has to be concurrent as much as possible, not only per test, but perhaps within test (between cases) if possible.
C. How hard is to add more tests / change tests / remove tests. Slow test (B) is one reason why ppl will not change/add tests, difficult to use framework or its complexity of use is another reason. People naturally will avoid this task or make it a minimum.
D. Being able to running (at least one test, not everything) on ready cluster of your choice for debugging and dev purposes.
E. Test framework complexity (amount of code, how flexible it is)

I fully agree: All e2e we don't need to run locally (so why running those in docker then and caring about portability?), however running one (on some cluster) is a common scenario (for me at least). I don't want to wait on all e2e tests and parse trillion logs that are parallelized (tests mix log lines) to check one failing test in detail 🤔 

At the end, for me it sounds like 40% of changes I would approve straightaway, 60% are not ideal at least for me and @TheSpiritXIII and you have opposite opinion. What we do about that? (: We could get others opinion too perhaps, but it's unfortunate the hard work was done by you before we got consensus.

[the remark this answers is not in the document]
Comment (bwplotka) [reply Makefile:208]: What e2e against GKE cluster allowed me: https://github.com/GoogleCloudPlatform/prometheus-engine/pull/783#issue-2100463945 (I don't need to run any kind cluster, can use console UI for manual double check)

[the remark this answers is not in the document]
Comment (pintohutch) [reply Makefile:208]: Sounds like we need to meet synchronously about this 😄 - IMO A, B, C, D, and E are all improvements with this chain of PRs...

[the remark this answers is not in the document]
Comment (pintohutch) [reply Makefile:208]: Or if @macxamin or @bernot-dev have opinions here please let me know (i.e. if you don't think we should be going this direction).

File: Makefile [hunk 3]
```diff
@@ -3,9 +3,24 @@ GOINSTR += $(notdir $(patsubst %/,%,$(dir $(shell find examples/instrumentation
 
 CLOUDSDK_CONFIG?=${HOME}/.config/gcloud
 PROJECT_ID?=$(shell gcloud config get-value core/project)
-GMP_CLUSTER?=gmp-test-cluster
+PROJECT_ID:=$(if $(PROJECT_ID),$(PROJECT_ID),gmp-project)
```
Comment (bernot-dev) [inline Makefile:6]: This line can just be `PROJECT_ID?=gmp-project`. The [question mark means it only proceeds if the variable is unassigned](https://www.gnu.org/software/make/manual/html_node/Setting.html#Setting-Variables), as above.

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (pintohutch) [reply Makefile:208]: Hey @bwplotka @TheSpiritXIII - are we good here after our sync?

**Review by TheSpiritXIII — APPROVED** [review]
LGTM! Just a small nit: look at Adam's comment. Thank you!

**Review by bwplotka — APPROVED** [review]

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply Makefile:208]: Yup 👍🏽 We did amazing consensus session and I am good

File: Makefile (hunk 3 above)
Comment (pintohutch) [reply Makefile:6]: Done
